nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -697,6 +1046,76 @@ async fn ssc_send_snapshot_request_undo(
     Ok(())
 }
 
+async fn ssc_call_pantry_detach_for_disk(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let params = sagactx.saga_params::<Params>()?;
+
+    let pantry_address = sagactx.lookup::<SocketAddrV6>("pantry_address")?;
+
+    let endpoint = format!("http://{}", pantry_address);
+
+    info!(
+        log,
+        "sending detach for disk {} to endpoint {}", params.disk_id, endpoint,
+    );
+
+    let client = crucible_pantry_client::Client::new(&endpoint);
+
+    retry_until_known_result!(log, {
+        client.detach(&params.disk_id.to_string())
+    });
+
+    Ok(())
+}
+
+async fn ssc_detach_disk_from_pantry(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let (.., authz_disk, db_disk) =
+        LookupPath::new(&opctx, &osagactx.datastore())
+            .disk_id(params.disk_id)
+            .fetch_for(authz::Action::Modify)
+            .await
+            .map_err(ActionError::action_failed)?;
+
+    match db_disk.state().into() {
+        external::DiskState::Maintenance => {
+            info!(
+                log,
+                "setting disk {} state from maintenance to detached",
+                params.disk_id
+            );
+
+            osagactx
+                .datastore()
+                .disk_update_runtime(
+                    &opctx,
+                    &authz_disk,
+                    &db_disk.runtime().detach(),
+                )
+                .await
+                .map_err(ActionError::action_failed)?;
+        }
+
+        external::DiskState::Detached => {
+            info!(log, "disk {} already detached", params.disk_id);
+        }
+
+        _ => {
+            warn!(log, "disk is in state {:?}", db_disk.state());
+        }
```

Comment:
If two concurrent operations are trying to bump a disk's generation number from N to N + 1, then only one can pass. If we're at this node in the saga, then this operation succeeded and the history of this disk's generation number and state is as follows:

```
Generation number | State
------------------|------------
            N + 0 | Detached
            N + 1 | Maintenance
```

Any attempt to bump the generation number of the disk from N + 1 to N + 2 will succeed if there are no other checks, and we would see two concurrent operations on the same disk. To make this situation impossible, no operation should begin on any disk in the maintenance state.

In the face of saga retries, let's say that this saga node executed fully but crashed before it could be committed to the database as finished. Then, the current state of the disk is "Detached" with the generation number N + 2. When this saga node is re-executed, there are two possibilities:

1. The disk is still "Detached" at generation number N + 2
2. The disk is in some unknown state at generation number N + 3

With 2, another operation started *after* the saga node executed but crashed, and *before* it was re-executed. This is a bug if there are further saga nodes to execute in the saga, which in this case there are. This call to `disk_update_runtime` has to be the last saga node that gets executed, which is done in 4c81a57e.